Automatically engage live-ish resize when available

hortont424 opened the pull request:
#### 202b4697ed2f7d0866688cb7439619110414e53a
<pre>
Automatically engage live-ish resize when available
<a href="https://bugs.webkit.org/show_bug.cgi?id=242534">https://bugs.webkit.org/show_bug.cgi?id=242534</a>

Reviewed by Wenson Hsieh.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
Add a preference for automatically beginning and ending live resize based
on system notifications.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _restorePageScrollPosition:scrollOrigin:previousObscuredInset:scale:]):
(-[WKWebView _restorePageStateToUnobscuredCenter:scale:]):
(-[WKWebView _scrollToContentScrollPosition:scrollOrigin:animated:]):
(-[WKWebView _shouldDeferGeometryUpdates]):
(-[WKWebView _updateVisibleContentRects]):
(-[WKWebView _setAvoidsUnsafeArea:]):
(-[WKWebView _setViewLayoutSizeOverride:]):
(-[WKWebView _setMinimumUnobscuredSizeOverride:]):
(-[WKWebView _setMaximumUnobscuredSizeOverride:]):
(-[WKWebView _setInterfaceOrientationOverride:]):
(-[WKWebView _snapshotRectAfterScreenUpdates:rectInViewCoordinates:intoImageOfWidth:completionHandler:]):
Factor `_dynamicViewportUpdateMode` checks out of callers, and into `_shouldDeferGeometryUpdates`.
Also defer geometry updates when we&apos;re in a live-ish resize, not just animated resize.

(-[WKWebView _beginAutomaticLiveResizeIfNeeded]):
Try to infer when we&apos;re in a live resize (currently in a somewhat roundabout way;
to be improved as better mechanisms become avilable), and call _beginLiveResize
and _endLiveResize automatically.

(-[WKWebView didMoveToWindow]):
(-[WKWebView _destroyEndLiveResizeObserver]):
Tear down any active live resize and end-observer when we move between
windows (or out of the window).

(-[WKWebView _frameOrBoundsChanged]):
(-[WKWebView _didCompleteAnimatedResize]):
(-[WKWebView _didStopDeferringGeometryUpdates]):
Factor `_didStopDeferringGeometryUpdates`, which sends all of the &quot;I stopped deferring
geometry updates, so now send the current state&quot; messages, out of _didCompleteAnimatedResize,
and call it from _endLiveResize, so that we always end up at the correct final state.

(-[WKWebView _beginLiveResize]):
(-[WKWebView _endLiveResize]):
Move these down to the internal category. Add some guards so we&apos;ll never start
an animated resize during live resize, or vice versa. Adopt `_didStopDeferringGeometryUpdates`
instead of randomly calling just `_frameOrBoundsChanged`.

(-[WKWebView _beginAnimatedResizeWithUpdates:]):

Canonical link: <a href="https://commits.webkit.org/252307@main">https://commits.webkit.org/252307@main</a>
</pre>

whsieh commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -2300,6 +2335,15 @@ - (UIEdgeInsets)currentlyVisibleContentInsetsWithScale:(CGFloat)scaleFactor obsc
+        return YES;
+    if (_liveResizeParameters)
+        return YES;
+    return NO;
```
Nit - would this be clearer as just `_liveResizeParameters || _dynamicViewportUpdateMode != WebKit::DynamicViewportUpdateMode::NotResizing`?

hortont424 replied:
Sure :)

whsieh commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -2014,6 +2014,37 @@ - (BOOL)_updateScrollViewContentInsetsIfNecessary
+
+    [self _beginLiveResize];
+    
+    _endLiveResizeNotificationObserver = [[NSNotificationCenter defaultCenter] addObserverForName:UISceneDidActivateNotification object:self.window.windowScene queue:NSOperationQueue.mainQueue usingBlock:makeBlockPtr([weakSelf = WeakObjCPtr<WKWebView>(self)] (NSNotification *note) {
```
Nit - you could probably omit the `note` parameter too, since it's unused in this block.

hortont424 committed:
Automatically engage live-ish resize when available
https://bugs.webkit.org/show_bug.cgi?id=242534

Reviewed by Wenson Hsieh.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
Add a preference for automatically beginning and ending live resize based
on system notifications.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _restorePageScrollPosition:scrollOrigin:previousObscuredInset:scale:]):
(-[WKWebView _restorePageStateToUnobscuredCenter:scale:]):
(-[WKWebView _scrollToContentScrollPosition:scrollOrigin:animated:]):
(-[WKWebView _shouldDeferGeometryUpdates]):
(-[WKWebView _updateVisibleContentRects]):
(-[WKWebView _setAvoidsUnsafeArea:]):
(-[WKWebView _setViewLayoutSizeOverride:]):
(-[WKWebView _setMinimumUnobscuredSizeOverride:]):
(-[WKWebView _setMaximumUnobscuredSizeOverride:]):
(-[WKWebView _setInterfaceOrientationOverride:]):
(-[WKWebView _snapshotRectAfterScreenUpdates:rectInViewCoordinates:intoImageOfWidth:completionHandler:]):
Factor `_dynamicViewportUpdateMode` checks out of callers, and into `_shouldDeferGeometryUpdates`.
Also defer geometry updates when we're in a live-ish resize, not just animated resize.

(-[WKWebView _beginAutomaticLiveResizeIfNeeded]):
Try to infer when we're in a live resize (currently in a somewhat roundabout way;
to be improved as better mechanisms become avilable), and call _beginLiveResize
and _endLiveResize automatically.

(-[WKWebView didMoveToWindow]):
(-[WKWebView _destroyEndLiveResizeObserver]):
Tear down any active live resize and end-observer when we move between
windows (or out of the window).

(-[WKWebView _frameOrBoundsChanged]):
(-[WKWebView _didCompleteAnimatedResize]):
(-[WKWebView _didStopDeferringGeometryUpdates]):
Factor `_didStopDeferringGeometryUpdates`, which sends all of the "I stopped deferring
geometry updates, so now send the current state" messages, out of _didCompleteAnimatedResize,
and call it from _endLiveResize, so that we always end up at the correct final state.

(-[WKWebView _beginLiveResize]):
(-[WKWebView _endLiveResize]):
Move these down to the internal category. Add some guards so we'll never start
an animated resize during live resize, or vice versa. Adopt `_didStopDeferringGeometryUpdates`
instead of randomly calling just `_frameOrBoundsChanged`.

(-[WKWebView _beginAnimatedResizeWithUpdates:]):

Canonical link: https://commits.webkit.org/252307@main
